### PR TITLE
[4.0] Multilingual status module improvements

### PR DESCRIPTION
**Pull request by infograf768**
Replaces https://github.com/joomla/joomla-cms/pull/24062 which had conflicts

This PR adds a Notice and displays the language concerned when the Content Language for that language has been deleted and remains in the site an orphan home page while the site language is published.
It also adds a Notice when a Content Language is in the trash.

It also reinstate all states of a content language to display the correct icons.
published = 1
published = 0
published = -2 (i.e. trash) and for this one displays the trash icon

It also modifies all the queries in the Helper to use quotename()

See more below

### Testing Instructions
Install a multilingual site with all languages available to install
French fr-FR
German de-DE
Persian fa-IR

Then use the Cpanel Multilingual Sample Data module to create a basic.

First test:
Display the Content languages Manager
Trash AND delete the fa-IR Persian content language.

You should get this when displaying the multilingual status module

screen shot 2019-03-02 at 18 25 09

Then unpublish the French fr-FR Content Language.
And trash (do NOT delete) the German de-DE Content language.
I.e. Manager should look like this
screen shot 2019-03-02 at 18 28 10

You should get this when displaying the multilingual status module
screen shot 2019-03-02 at 18 44 18

Unpublish also the fr-FR Home page to get a new notice.
Deleted Content Language error now displays the notice as last.
see https://github.com/joomla/joomla-cms/pull/24062#issuecomment-469004768

Added new Notice when both Content Language and Home page are unpublished.
We should now get in that case:

screen shot 2019-03-03 at 10 22 15

 Added notice when both content language and home page missing. added trash icon when home page is in the trash.

screen shot 2019-03-04 at 12 40 18

Added warning icon when home page does not exist.
screen shot 2019-03-04 at 12 56 37

### Reminder
It is OK to have an Unpublished Content Language (as French is above) as it lets the site owner prepare a new Language for the multilingual site.

The screenshots were taken in March. They will now rather look like this:
Screen Shot 2019-11-30 at 08 44 07

@Quy
@alikon 
Please test and specially look at the queries as I did my best but there could be improvements there.

### CODE REVIEW

File: administrator/components/com_languages/Helper/MultilangstatusHelper.php [hunk 1]
```diff
@@ -95,20 +95,20 @@ public static function getStatus()
...
-			->join('LEFT', '#__extensions  AS e ON e.element = a.lang_code')
-			->where('e.client_id = 0')
-			->where('e.enabled = 1')
-			->where('e.state = 0');
+		$query->select($db->quoteName('l.home', 'home'))
+			->select($db->quoteName('l.published', 'home_published'))
+			->join('LEFT', $db->quoteName('#__menu', 'l') . ' ON ' . $db->quoteName('l.language') . ' = ' . $db->quoteName('a.lang_code') . ' AND ' . $db->quoteName('l.home') . ' = 1  AND ' . $db->quoteName('l.language') . ' <> \'*\'')
+			->select($db->quoteName('e.enabled', 'enabled'))
+			->select($db->quoteName('e.element', 'element'))
+			->join('LEFT', $db->quoteName('#__extensions', 'e') . ' ON ' . $db->quoteName('e.element') . ' = ' . $db->quoteName('a.lang_code'))
```
Comment (Quy) [inline administrator/components/com_languages/Helper/MultilangstatusHelper.php]:
```suggestion
			->join('LEFT', $db->quoteName('#__extensions', 'e'), $db->quoteName('e.element') . ' = ' . $db->quoteName('a.lang_code'))
```

File: administrator/components/com_languages/Helper/MultilangstatusHelper.php [hunk 2]
```diff
@@ -36,9 +36,9 @@ public static function getHomes()
 		$query = $db->getQuery(true)
 			->select('COUNT(*)')
 			->from($db->quoteName('#__menu'))
-			->where('home = 1')
-			->where('published = 1')
-			->where('client_id = 0');
+			->where($db->quoteName('home') . ' = 1')
```
Comment (Quy) [inline administrator/components/com_languages/Helper/MultilangstatusHelper.php]: Combine like this: https://github.com/joomla/joomla-cms/pull/27107/files#diff-f86013e28841dc37d7a3624d3bfd2cecR1069

File: administrator/components/com_languages/Helper/MultilangstatusHelper.php [hunk 3]
```diff
@@ -74,9 +74,9 @@ public static function getContentlangs()
 		// Check for published Content Languages.
 		$db = Factory::getDbo();
 		$query = $db->getQuery(true)
-			->select('a.lang_code AS lang_code')
-			->select('a.published AS published')
-			->from('#__languages AS a');
+			->select($db->quoteName('a.lang_code', 'lang_code'))
```
Comment (Quy) [inline administrator/components/com_languages/Helper/MultilangstatusHelper.php]: Combine like this: https://github.com/joomla/joomla-cms/pull/27107/files#diff-f86013e28841dc37d7a3624d3bfd2cecR501

**Comment by infograf768** [comment]
@Quy 
Also did other combined queries, as much as I could do.
Could be missing some in the getStatus() method but I honestly think it would make the queries there more confusing than benefiting as we have multiple select.
Also I am not aware if we need or not to also change line 130 join and if we do as on line 123, if we should keep the `' AND '` as is.

May I suggest now to test the results of this PR rather than concentrating on the queries format as that type of format is absent from the majority of 4.0 code.

File: administrator/components/com_languages/Helper/MultilangstatusHelper.php [hunk 4]
```diff
@@ -199,10 +230,14 @@ public static function getDefaultHomeModule()
 		$query = $db->getQuery(true)
 			->select($db->quoteName('menutype'))
 			->from($db->quoteName('#__menu'))
-			->where($db->quoteName('home') . ' = ' . $db->quote('1'))
-			->where($db->quoteName('published') . ' = ' . $db->quote('1'))
-			->where($db->quoteName('client_id') . ' = ' . $db->quote('0'))
-			->where($db->quoteName('language') . ' = ' . $db->quote('*'));
+			->where(
+				[
+					$db->quoteName('home') . ' = ' . $db->quote('1'),
```
Comment (Quy) [inline administrator/components/com_languages/Helper/MultilangstatusHelper.php]: Don't apply `$db->quote()` to lines 235-237 as these are integer columns.

File: administrator/components/com_languages/Helper/MultilangstatusHelper.php [hunk 5]
```diff
@@ -212,9 +247,13 @@ public static function getDefaultHomeModule()
 		$query->clear()
 			->select($db->quoteName('title'))
 			->from($db->quoteName('#__modules'))
-			->where($db->quoteName('module') . ' = ' . $db->quote('mod_menu'))
-			->where($db->quoteName('published') . ' = ' . $db->quote('1'))
-			->where($db->quoteName('client_id') . ' = ' . $db->quote('0'));
+			->where(
+				[
+					$db->quoteName('module') . ' = ' . $db->quote('mod_menu'),
+					$db->quoteName('published') . ' = ' . $db->quote('1'),
```
Comment (Quy) [inline administrator/components/com_languages/Helper/MultilangstatusHelper.php]: Don't apply `$db->quote()` to lines 253-254 as these are integer columns.

File: administrator/components/com_languages/Helper/MultilangstatusHelper.php [hunk 6]
```diff
@@ -253,9 +292,13 @@ public static function getModule($moduleName, $instanceTitle = null)
 		$query = $db->getQuery(true)
 			->select('id, title, module, position, content, showtitle, params')
 			->from($db->quoteName('#__modules'))
-			->where($db->quoteName('module') . ' = ' . $db->quote($moduleName))
-			->where($db->quoteName('published') . ' = ' . $db->quote('1'))
-			->where($db->quoteName('client_id') . ' = ' . $db->quote('0'));
+			->where(
+				[
+					$db->quoteName('module') . ' = ' . $db->quote($moduleName),
+					$db->quoteName('published') . ' = ' . $db->quote('1'),
```
Comment (Quy) [inline administrator/components/com_languages/Helper/MultilangstatusHelper.php]: Don't apply $db->quote() to lines 298-299 as these are integer columns.

**Comment by infograf768** [comment]
Done.

File: administrator/components/com_languages/Helper/MultilangstatusHelper.php (hunk 4 above)
Comment (Quy) [inline administrator/components/com_languages/Helper/MultilangstatusHelper.php]:
```suggestion
					$db->quoteName('home') . ' = 1',
```

Comment (infograf768) [reply administrator/components/com_languages/Helper/MultilangstatusHelper.php]: done

**Comment by Quy** [comment]
These are redundant.

> The de-DE Content Language is trashed. It will not display on the site. 

> The Content Language de-DE is trashed.

**Comment by infograf768** [comment]
> These are redundant.

Good find! Corrected.

File: administrator/components/com_languages/tmpl/multilangstatus/default.php [hunk 7]
```diff
@@ -188,60 +221,58 @@
...
-					<?php if ($status->home_language) : ?>
-							<td class="text-center">
+						<td class="text-center">
+							<?php if ($status->home_published == 1) : ?>
 								<span class="fa fa-check" aria-hidden="true"></span>
 								<span class="sr-only"><?php echo Text::_('JYES'); ?></span>
-							</td>
-					<?php else : ?>
-							<td class="text-center">
+							<?php elseif ($status->home_published === 0) : ?>
```
Comment (Quy) [inline administrator/components/com_languages/tmpl/multilangstatus/default.php]: Strict comparison is done here, but not with the others.

Comment (SharkyKZ) [reply administrator/components/com_languages/tmpl/multilangstatus/default.php]: It shouldn't be here unless values are explicitly cast as integers.

Comment (infograf768) [reply administrator/components/com_languages/tmpl/multilangstatus/default.php]: Done.

**Comment by infograf768** [comment]
Still discussing with Mig about some display and content of messages.
Changes would be cosmetic.
So, can be tested but still WIP.

**Comment by brianteeman** [comment]
just a question. Is there a reason that we load this page in a popup and not a regular page

**Comment by infograf768** [comment]
> Is there a reason that we load this page in a popup and not a regular page

The reason was that it should be available at any time from any page of admin as some changes can have unwanted consequences on multilingual.

**Comment by brianteeman** [comment]
If it was created like the postinstallation messages wouldn't that satisfy that.

the reason is that the amount of information in the popup is getting quite large and it will be easier to display etc on a full page

**Comment by infograf768** [comment]
It has absolutely nothing to do with postinstallation messages and comparing these is like comparing an apple to a bicycle.

Loading it like it is done for Postinstallation messages, which are usually only loaded once and can be hidden once read, is not what we want for this module as it would force to navigate again to where it was first needed by the user. 
In 4.0, it may be a pain as, when using the default Preset menu, one may have to go through the System menu, etc.

Yes, when many aspects are wrong, it can become quite large. It is only because of the possible warnings. These important errors can be easily corrected grace to the informations provided which soon reduces the size. Scroll bars are there to be used if necessary.

This is working fine as is and makes it easier to inform users while preventing useless endless navigation.

Usability is the key word here.

**Comment by brianteeman** [comment]
> It has absolutely nothing to do with postinstallation messages and comparing these is like comparing an apple to a bicycle.

I was referring to the location of the link

> Loading it like it is done for Postinstallation messages, which are usually only loaded once and can be hidden once read, is not what we want for this module as it would force to navigate again to where it was first needed by the user.

Not if you set a return value

> Yes, when many aspects are wrong, it can become quite large. It is only because of the possible warnings. These important errors can be easily corrected grace to the informations provided which soon reduces the size. Scroll bars are there to be used if necessary.

My suggestion removes the need for the scrollbar

> Usability is the key word here.

Exactly which is why I make the suggestion

**Comment by infograf768** [comment]
One is totally free to propose, once this PR is completed an d merged, another way to display these informations. 
This is obviously not the purpose of this PR.

**Comment by brianteeman** [comment]
Now that I know there is no technical reason I will do

File: administrator/components/com_languages/Helper/MultilangstatusHelper.php [hunk 8]
```diff
@@ -128,37 +147,49 @@ public static function getContacts()
...
 		// Get the number of languages for the contact
 		$slang = $db->getQuery(true)
 			->select('count(distinct(l.lang_code))')
-			->from('#__languages as l')
-			->join('LEFT', '#__contact_details AS cd ON cd.language=l.lang_code')
-			->where('cd.user_id=u.id')
-			->where('cd.published=1')
-			->where('l.published=1');
+			->from($db->quoteName('#__languages', 'l'))
+			->join('LEFT', $db->quoteName('#__contact_details', 'cd') . ' ON ' . $db->quoteName('cd.language') . '= ' . $db->quoteName('l.lang_code'))
```
Comment (Quy) [inline administrator/components/com_languages/Helper/MultilangstatusHelper.php]:
```suggestion
			->join('LEFT', $db->quoteName('#__contact_details', 'cd') . ' ON ' . $db->quoteName('cd.language') . ' = ' . $db->quoteName('l.lang_code'))
```

Comment (Quy) [inline administrator/components/com_languages/Helper/MultilangstatusHelper.php]:
```suggestion
			->join('LEFT', $db->quoteName('#__contact_details', 'cd'), $db->quoteName('cd.language') . ' = ' . $db->quoteName('l.lang_code'))
```

Comment (infograf768) [reply administrator/components/com_languages/Helper/MultilangstatusHelper.php]: Done, and as well for the 2 following queries.

File: administrator/components/com_languages/Helper/MultilangstatusHelper.php [hunk 9]
```diff
@@ -128,37 +147,49 @@ public static function getContacts()
...
 
 		// Get the contacts
 		$query = $db->getQuery(true)
 			->select('u.name, (' . $alang . ') as alang, (' . $slang . ') as slang, (' . $mlang . ') as mlang')
-			->from('#__users AS u')
-			->join('LEFT', '#__contact_details AS cd ON cd.user_id=u.id')
-			->where('EXISTS (SELECT 1 from #__content as c where  c.created_by=u.id)')
+			->from($db->quoteName('#__users', 'u'))
+			->join('LEFT', $db->quoteName('#__contact_details', 'cd'), $db->quoteName('cd.user_id') . ' = ' . $db->quoteName('u.id'))
+			->where('EXISTS (SELECT 1 from' . $db->quoteName('#__content', 'c') . ' where ' . $db->quoteName('c.created_by') . ' = ' . $db->quoteName('u.id') . ')')
```
Comment (Quy) [inline administrator/components/com_languages/Helper/MultilangstatusHelper.php]:
```suggestion
			->where('EXISTS (SELECT 1 FROM ' . $db->quoteName('#__content', 'c') . ' WHERE ' . $db->quoteName('c.created_by') . ' = ' . $db->quoteName('u.id') . ')')
```

Comment (infograf768) [reply administrator/components/com_languages/Helper/MultilangstatusHelper.php]: done

File: administrator/components/com_languages/Helper/MultilangstatusHelper.php [hunk 10]
```diff
@@ -128,37 +147,49 @@ public static function getContacts()
...
 
 		// Get the contacts
 		$query = $db->getQuery(true)
 			->select('u.name, (' . $alang . ') as alang, (' . $slang . ') as slang, (' . $mlang . ') as mlang')
-			->from('#__users AS u')
-			->join('LEFT', '#__contact_details AS cd ON cd.user_id=u.id')
-			->where('EXISTS (SELECT 1 from #__content as c where  c.created_by=u.id)')
+			->from($db->quoteName('#__users', 'u'))
+			->join('LEFT', $db->quoteName('#__contact_details', 'cd'), $db->quoteName('cd.user_id') . ' = ' . $db->quoteName('u.id'))
+			->where('EXISTS (SELECT 1 FROM' . $db->quoteName('#__content', 'c') . ' WHERE ' . $db->quoteName('c.created_by') . ' = ' . $db->quoteName('u.id') . ')')
```
Comment (Quy) [inline administrator/components/com_languages/Helper/MultilangstatusHelper.php]:
```suggestion
			->where('EXISTS (SELECT 1 FROM ' . $db->quoteName('#__content', 'c') . ' WHERE ' . $db->quoteName('c.created_by') . ' = ' . $db->quoteName('u.id') . ')')
```

Comment (infograf768) [reply administrator/components/com_languages/Helper/MultilangstatusHelper.php]: done

File: administrator/components/com_languages/tmpl/multilangstatus/default.php [hunk 11]
```diff
@@ -60,7 +75,7 @@
 			</div>
 		<?php endif; ?>
 		<?php foreach ($this->contentlangs as $contentlang) : ?>
-			<?php if (array_key_exists($contentlang->lang_code, $this->homepages) && (!array_key_exists($contentlang->lang_code, $this->site_langs) || !$contentlang->published)) : ?>
+			<?php if (array_key_exists($contentlang->lang_code, $this->homepages) && (!array_key_exists($contentlang->lang_code, $this->site_langs) || $contentlang->published !== 1)) : ?>
```
Comment (Quy) [inline administrator/components/com_languages/tmpl/multilangstatus/default.php]:
```suggestion
			<?php if (array_key_exists($contentlang->lang_code, $this->homepages) && (!array_key_exists($contentlang->lang_code, $this->site_langs) || $contentlang->published != 1)) : ?>
```

Comment (infograf768) [reply administrator/components/com_languages/tmpl/multilangstatus/default.php]: Done

**Comment by infograf768** [comment]
Added a new Error when the Default Site Language has no published Content Language.
As this would totally break the site, It is displayed as an Error on top of other warnings.
To test, unpublish, trash or delete the Content Language related to the Site Default Language.
You should get this:

Screen Shot 2019-12-10 at 08 12 52

The second part of the new message "If not corrected, the multilingual feature will not work!"
could be improved. Proposals welcome!

**Comment by infograf768** [comment]
Changed new string to
`COM_LANGUAGES_MULTILANGSTATUS_ERROR_DEFAULT_CONTENT_LANGUAGE="The Content Language for the <strong>%s</strong> default Site Language is unpublished or does not exist. <strong>If not corrected, the multilingual feature may not function properly or the site may not be reachable!</strong>"`

Screen Shot 2019-12-10 at 10 48 14


Thanks Mig.

I guess it is now ready to go, as a module.

**Comment by imanickam** [comment]
I have tested this item :white_check_mark: successfully on 605b87eadf06c6c1e482c861b799ab2d71617ab5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27184">issues.joomla.org/tracker/joomla-cms/27184</a>.</sub>

File: administrator/components/com_languages/tmpl/multilangstatus/default.php [hunk 12]
```diff
@@ -188,60 +230,58 @@
...
-							<span class="icon-trash" aria-hidden="true"></span>
-							<span class="sr-only"><?php echo Text::_('WARNING'); ?></span>
-						</td>
-					<?php else : ?>
-							<td class="text-center">
+							<?php elseif ($status->lang_code && $status->published == 0) : ?>
+								<span class="fa fa-times" aria-hidden="true"></span>
+								<span class="sr-only"><?php echo Text::_('WARNING'); ?></span>
+							<?php elseif ($status->lang_code && $status->published == -2) : ?>
+								<span class="icon-trash" aria-hidden="true"></span>
```
Comment (Quy) [inline administrator/components/com_languages/tmpl/multilangstatus/default.php]: The trash icon is not centered due to the right margin.

Comment (infograf768) [reply administrator/components/com_languages/tmpl/multilangstatus/default.php]: Corrected.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 0cc3ab884cb78197fe758ce355a9f1214f0fdc0d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27184">issues.joomla.org/tracker/joomla-cms/27184</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27184">issues.joomla.org/tracker/joomla-cms/27184</a>.</sub>

**Comment by HLeithner** [comment]
Thanks

**Comment by infograf768** [comment]
@brianteeman 
I can make a new patch implementing what you wanted, with return  when Close to the place where the Multilingual status was clicked.
multilang_full page

Is that what you want?

**Comment by brianteeman** [comment]
That is almost what I was talking about - give me one sec

**Comment by brianteeman** [comment]
The url would be 
index.php?option=com_languages&view=multilangstatus&return=xxxxxxxxxxx

without the template=component so that it looks to be part of the site

image

**Comment by infograf768** [comment]
I will let you do it. I found some issues here doing it like this when switching to other menus without closing first the multilingual status module.

EDIT: also it lets click again on the module icon and we get into an endless loop as the redirect would now be the multilangstatus module path itself.

**Comment by infograf768** [comment]
In any case the return is obtained by
`$redirect = Uri::getInstance()->toString();` in the module default.php

**Comment by brianteeman** [comment]
ok - I will put it on my todo

**Pull request merged**